src/iosrv/drpc_ras.c

```diff
@@ -0,0 +1,283 @@
+
+static void
+free_ras(Shared__RASEvent *evt)
+{
+	if (evt->hostname)
+		D_FREE(evt->hostname);
+	if (evt->timestamp)
+		D_FREE(evt->timestamp);
+}
+
+static int
+init_ras(ras_event_t id, char *msg, ras_type_t type, ras_sev_t sev, char *hid,
+	 d_rank_t *rank, char *jid, uuid_t *puuid, uuid_t *cuuid,
+	 daos_obj_id_t *oid, char *cop, Shared__RASEvent *evt)
+{
+	FILE		*stream;
+	char		*buf;
+	size_t		 len;
+	struct timeval	 tv;
+	struct tm	*tm;
+	struct utsname	 uts;
+	int		 rc;
+
+	stream = open_memstream(&buf, &len);
+	if (!stream)
+		return -DER_NOMEM;
+
+	/* populate mandatory RAS fields */
+
+	evt->id = (uint32_t)id;
+	D_FPRINTF(stream, " id: [%s]", ras_event2str(id));
+
+	(void)gettimeofday(&tv, 0);
+	tm = localtime(&tv.tv_sec);
+	if (tm) {
+		D_ALLOC(evt->timestamp, DAOS_RAS_STR_FIELD_SIZE);
+		if (!evt->timestamp)
+			D_GOTO(out_fail, rc = -DER_NOMEM);
+		D_ASPRINTF(evt->timestamp,
+			   "%04d/%02d/%02d-%02d:%02d:%02d.%02ld",
+			   tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
+			   tm->tm_hour, tm->tm_min, tm->tm_sec,
+			   (long)tv.tv_usec / 10000);
+		D_FPRINTF(stream, " ts: [%s]", evt->timestamp);
+	}
+
+	/** XXX: nodename should be fetched only once and not on every call */
+	D_ALLOC(evt->hostname, DAOS_RAS_STR_FIELD_SIZE);
+	if (!evt->hostname)
+		D_GOTO(out_fail, rc = -DER_NOMEM);
+	(void)uname(&uts);
+	D_ASPRINTF(evt->hostname, "%s", uts.nodename);
+	D_FPRINTF(stream, " host: [%s]", evt->hostname);
+
+	evt->type = (uint32_t)type;
+	evt->severity = (uint32_t)sev;
+	D_FPRINTF(stream, " type: [%s] sev: [%s]", ras_type2str(type),
+		  ras_sev2str(sev));
+
+	if (!msg) {
+		D_ERROR("missing msg parameter\n");
+		D_GOTO(out_fail, rc = -DER_INVAL);
+	}
+	evt->msg = msg;
+	D_FPRINTF(stream, " msg: [%s]", msg);
+
+	/* populate optional RAS fields */
+
+	if (hid) {
+		evt->hw_id = hid;
+		D_FPRINTF(stream, " hwid: [%s]", hid);
+	}
+
+	if (rank) {
+		evt->rank = (uint32_t)*rank;
+		D_FPRINTF(stream, " rank: [%u]", *rank);
+	}
+
+	if (jid) {
+		evt->job_id = jid;
+		D_FPRINTF(stream, " jobid: [%s]", jid);
+	}
+
+	if (puuid && !uuid_is_null(*puuid)) {
+		D_ALLOC(evt->pool_uuid, DAOS_UUID_STR_SIZE);
+		if (!evt->pool_uuid)
+			D_GOTO(out_fail, rc = -DER_NOMEM);
+		D_ASPRINTF(evt->pool_uuid, DF_UUIDF, DP_UUID(*puuid));
+		D_FPRINTF(stream, " puuid: ["DF_UUIDF"]", DP_UUID(*puuid));
+	}
+
+	if (cuuid && !uuid_is_null(*cuuid)) {
+		D_ALLOC(evt->cont_uuid, DAOS_UUID_STR_SIZE);
+		if (!evt->cont_uuid)
+			D_GOTO(out_fail, rc = -DER_NOMEM);
+		D_ASPRINTF(evt->cont_uuid, DF_UUIDF, DP_UUID(*cuuid));
+		D_FPRINTF(stream, " cuuid: ["DF_UUIDF"]", DP_UUID(*cuuid));
+	}
+
+	if (oid) {
+		D_ALLOC(evt->obj_id, DAOS_RAS_STR_FIELD_SIZE);
+		if (!evt->obj_id)
+			D_GOTO(out_fail, rc = -DER_NOMEM);
+		D_ASPRINTF(evt->obj_id, DF_OID, DP_OID(*oid));
+		D_FPRINTF(stream, " oid: ["DF_OID"]", DP_OID(*oid));
+	}
+
+	if (cop) {
+		evt->ctl_op = cop;
+		D_FPRINTF(stream, " control_op: [%s]", cop);
+	}
+
+	fclose(stream);
+	D_DEBUG(DB_MGMT, "&&& RAS EVENT%s", buf);
```

Comment:
Wouldn't it be better to have init_ras() just populating the evt and then have a function that print the ras event for the debug logs from the evt?

Reply:
ok